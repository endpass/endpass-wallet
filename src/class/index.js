export { providerFactory } from './provider';
export { default as Transaction } from './Transaction';
export {
  default as TransactionFactory,
} from './transaction/TransactionFactory';
export { default as EventEmitter } from './EventEmitter';
export { NotificationError } from './error';
export { default as Address } from './address';
export { default as Wallet } from './wallet';
export { ENSResolver } from './ens';
export { Token } from './Token';
export { default as ERC20Token } from './erc20';
export { default as TrezorWallet } from './hardwareWallet/TrezorWallet';
export { default as LedgerWallet } from './hardwareWallet/LedgerWallet';
