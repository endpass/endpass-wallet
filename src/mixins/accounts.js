// Mixin for common methods for Ethereum accounts
import web3 from 'web3';

const { fromWei } = web3.utils;

export default {
  computed: {
    address() {
      return this.$store.state.accounts.address;
    },

    balance() {
      return this.$store.state.accounts.balance === null
        ? null
        : fromWei(this.$store.getters['accounts/balance']);
    },
  },
};
