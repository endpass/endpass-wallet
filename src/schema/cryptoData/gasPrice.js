export const cryptoDataGasPrice = {
  type: 'object',
  additionalProperties: false,
  required: ['low', 'medium', 'high'],
  properties: {
    low: {
      type: 'number',
    },
    medium: {
      type: 'number',
    },
    high: {
      type: 'number',
    },
  },
};
