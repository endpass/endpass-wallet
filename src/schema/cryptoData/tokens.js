export const cryptoDataToken = {
  type: 'object',
  required: ['name', 'symbol', 'address'],
  properties: {
    address: {
      type: 'string',
      pattern: '^0x[0-9A-Fa-f]+$',
    },
    name: {
      type: 'string',
      minLength: 1,
    },
    symbol: {
      type: 'string',
      minLength: 1,
    },
    balance: {
      type: 'string',
      pattern: '[0-9]+',
      default: '0',
    },
    price: {
      type: 'boolean',
      default: false,
    },
  },
};

export const cryptoDataTokens = {
  type: 'array',
  items: cryptoDataToken,
};
