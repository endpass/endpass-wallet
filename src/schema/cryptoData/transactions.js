export const cryptoDataTransaction = {
  type: 'object',
  required: ['timestamp', 'from', 'to', 'hash', 'value'],
  properties: {
    data: {
      type: 'string',
      pattern: '^0x[a-zA-Z0-9]+',
    },
    from: {
      type: 'string',
      pattern: '^0x[a-zA-Z0-9]{40,}',
    },
    hash: {
      type: 'string',
      pattern: '^0x[a-zA-Z0-9]{64,}',
    },
    timestamp: {
      type: 'number',
    },
    to: {
      type: 'string',
      pattern: '^0x[a-zA-Z0-9]{40,}',
    },
    // token: {
    //   type: 'object',
    //   required: ['name', 'symbol'],
    //   properties: {
    //     name: {
    //       type: 'string',
    //     },
    //     symbol: {
    //       type: 'string',
    //     },
    //     decimals: {
    //       type: 'number',
    //       default: 18,
    //     },
    //   },
    // },
    value: {
      type: 'number',
    },
  },
};

export const cryptoDataTransactions = {
  type: 'array',
  items: cryptoDataTransaction,
};
