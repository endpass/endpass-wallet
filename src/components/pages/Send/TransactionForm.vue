<template>
  <v-form
    id="sendEther"
    data-test="transaction-send-form"
    @submit="handleFormSubmit"
  >
    <div class="field">
      <label class="label">
        To
      </label>
      <account-chooser
        v-model="address"
        :disabled="isLoading || isEnsAddressLoading"
        :creatable="true"
        :width="35"
        :accounts="allowedToSendAddresses"
        placeholder="0x... or ENS"
        data-test="transaction-address-select"
      />
      <p
        v-if="isEnsTransaction && !ensError && !isEnsAddressLoading"
        class="help ellipsis"
      >
        Resolved ENS address: {{ transaction.to }}
      </p>
      <p
        v-if="ensError && !isEnsAddressLoading"
        class="help is-danger"
      >
        {{ ensError }}
      </p>
    </div>

    <transaction-amount-options
      v-model="transaction.value"
      :tokens-currencies="currentAccountTokensCurrencies"
      :current-token="transaction.tokenInfo"
      :fiat-currency="fiatCurrency"
      :eth-price="ethPrice"
      :balance="balance"
      :estimated-gas-cost="estimatedGasCost"
      :disabled="!address"
      :active-net="activeNet"
      :is-loading="isLoading"
      :show-fee="!!transaction.to"
      @change-token="changeTokenInfo"
    />

    <transaction-priority-options
      v-model="transaction.gasPrice"
      :is-loading="isLoadingGasPrice"
      :prices="prices"
    />

    <transaction-advanced-options
      :transaction="transaction"
      :current-token="transaction.tokenInfo"
      :is-loading="isLoading"
      @change="handleAdvancedChange"
    />

    <div class="field is-horizontal">
      <div class="field-label" />
      <div class="field-body">
        <v-button
          :disabled="!isSendAllowed"
          :loading="isLoading"
          class-name="is-success is-medium is-cta"
          data-test="transaction-send-button"
        >
          Send
        </v-button>
      </div>
    </div>
  </v-form>
</template>

<script>
import { debounce } from 'lodash';
import { mapGetters, mapState, mapActions } from 'vuex';
import { ENSResolver, Transaction } from '@/class';
import VForm from '@/components/ui/form/VForm.vue';
import VRadio from '@/components/ui/form/VRadio.vue';
import VSelect from '@/components/ui/form/VSelect';
import VInput from '@/components/ui/form/VInput.vue';
import VSpinner from '@/components/ui/VSpinner';
import VInputAddress from '@/components/ui/form/VInputAddress.vue';
import VButton from '@/components/ui/form/VButton.vue';
import AccountChooser from '@/components/AccountChooser';
import TransactionAdvancedOptions from './TransactionAdvancedOptions.vue';
import TransactionAmountOptions from './TransactionAmountOptions.vue';
import TransactionPriorityOptions from './TransactionPriorityOptions.vue';

export default {
  name: 'TransactionForm',

  inject: ['$validator'],

  props: {
    isLoading: {
      type: Boolean,
      default: false,
    },

    transaction: {
      type: Object,
      required: true,
    },
  },

  data: () => ({
    address: '',
    prices: null,
    estimatedGasCost: 0,
    ensError: null,
    isEnsAddressLoading: false,
    isLoadingGasPrice: true,
    isEstimationInProcess: false,
  }),

  computed: {
    ...mapState({
      balance: state => state.accounts.balance,
      activeNet: state => state.web3.activeNet,
      fiatCurrency: state => state.user.settings.fiatCurrency,
      ethPrice: state => state.price.price || 0,
      isSyncing: state => !!state.connectionStatus.isSyncing,
    }),
    ...mapGetters('transactions', ['allowedToSendAddresses']),
    ...mapGetters('tokens', [
      'currentAccountTokensCurrencies',
      'currentAccountTokenBySymbol',
    ]),

    isEnsTransaction() {
      return /^.+\.(eth|etc|test)$/.test(this.address);
    },

    isSendAllowed() {
      const {
        transaction,
        ensError,
        isSyncing,
        isEnsAddressLoading,
        isEstimationInProcess,
      } = this;

      return (
        transaction.to &&
        !isSyncing &&
        !ensError &&
        !isEnsAddressLoading &&
        !isEstimationInProcess
      );
    },
  },

  watch: {
    async address() {
      if (this.isEnsTransaction) {
        this.transaction.to = await this.resolveEnsAddress();
      } else if (!this.errors.has('address') && this.address) {
        this.ensError = null;
        this.transaction.to = this.address;
      }

      this.debouncedGasCostEstimation(1);
    },

    async activeNet(newValue, prevValue) {
      const isNetworkChanged = newValue.id !== prevValue.id;

      if (!isNetworkChanged) return;

      if (this.isEnsTransaction) {
        this.transaction.to = await this.resolveEnsAddress();
      }

      this.debouncedGasCostEstimation(2);
    },

    'transaction.tokenInfo': {
      handler() {
        this.debouncedGasCostEstimation(3);
      },
    },

    'transaction.gasPrice': {
      handler() {
        this.debouncedGasCostEstimation(4);
      },
    },

    'transaction.gasLimit': {
      handler() {
        this.debouncedGasCostEstimation(5);
      },
    },

    'transaction.to': {
      handler() {
        if (!this.transaction.to) {
          this.address = '';
        }
      },
    },
  },

  methods: {
    ...mapActions('gasPrice', ['getGasPrice']),

    changeTokenInfo(value) {
      if (value) {
        const tokenInfo = this.currentAccountTokenBySymbol(value);

        this.$set(this.transaction, 'tokenInfo', tokenInfo);
      } else {
        this.$set(this.transaction, 'tokenInfo', null);
      }
    },

    handleFormSubmit() {
      this.emitTransactionSubmit();
    },

    async resolveEnsAddress() {
      this.isEnsAddressLoading = true;

      try {
        const ensAddress = await ENSResolver.getAddress(this.address);

        this.ensError = null;

        return ensAddress;
      } catch (err) {
        this.ensError = `ENS ${this.address} can not be resolved.`;

        return '';
      } finally {
        this.isEnsAddressLoading = false;
      }
    },

    debouncedGasCostEstimation: debounce(function() {
      this.estimateGasCost();
    }, 500),

    async estimateGasCost() {
      if (!this.transaction.to) return;

      this.isEstimationInProcess = true;

      try {
        this.estimatedGasCost = await Transaction.getGasFullPrice(
          this.transaction,
        );
      } catch (err) {
        // TODO: проверить на реальном эфире, отправляется ли. Если да, при ошибке запрещать отправку
        console.log(err);

        const isContract = await Transaction.isTransactionToContract(
          this.transaction,
        );

        if (!isContract && err.message.includes('always failing transaction')) {
          this.ensError = 'Transaction will always fail, try other address.';
        }

        this.estimatedGasCost = 0;
      } finally {
        this.isEstimationInProcess = false;
      }
    },

    handleAdvancedChange(options) {
      Object.assign(this.transaction, options);
    },

    async loadGasPrice() {
      this.isLoadingGasPrice = true;

      this.prices = await this.getGasPrice();
      this.transaction.gasPrice = this.prices.medium.toString();

      this.isLoadingGasPrice = false;
    },

    emitTransactionSubmit() {
      this.$emit('submit', this.transaction);
    },
  },

  async created() {
    await this.loadGasPrice();
  },

  components: {
    VForm,
    VRadio,
    VSelect,
    VInput,
    VSpinner,
    VInputAddress,
    VButton,
    AccountChooser,
    TransactionAdvancedOptions,
    TransactionAmountOptions,
    TransactionPriorityOptions,
  },
};
</script>

<style>
</style>
