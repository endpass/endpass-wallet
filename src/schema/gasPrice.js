import { ajv } from '@/class/singleton';

export const gasPriceSchema = {
  properties: {
    low: {
      type: 'number',
      default: 0,
    },
    medium: {
      type: 'number',
      default: 0,
    },
    high: {
      type: 'number',
      default: 0,
    },
  },
};

export default {
  validateGasPrice: ajv.compile(gasPriceSchema),
};
