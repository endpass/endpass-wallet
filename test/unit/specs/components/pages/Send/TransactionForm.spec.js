import { set } from 'lodash';
import Vuex from 'vuex';
import VeeValidate from 'vee-validate';
import { createLocalVue } from '@vue/test-utils';
import UIComponents from '@endpass/ui';
import { Transaction } from '@/class';
import ENSResolver from '@/class/ens';
import validation from '@/validation';
import { wrapShallowMountFactory } from '@/testUtils';

import TransactionForm from '@/components/pages/Send/TransactionForm.vue';

import { transaction } from 'fixtures/transactions';
import { address } from 'fixtures/accounts';
import { token, tokens } from 'fixtures/tokens';
import { gasPrice } from 'fixtures/gasPrice';

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(validation);
localVue.use(VeeValidate);
localVue.use(UIComponents);

describe('Send â€“ TransactionForm', () => {
  let store;
  let wrapper;
  let wrapperFactory;

  const tokensGetters = {
    currentAccountTokensCurrencies: jest.fn(),
    currentAccountTokenBySymbol: () => jest.fn(),
  };
  const gasPriceActions = {
    getGasPrice: jest.fn().mockResolvedValue(gasPrice),
  };
  const transactionProp = {
    ...transaction,
    gasPrice: 0,
    tokenInfo: token,
  };
  const defaultStore = {
    state: {
      accounts: {
        balance: '1000000',
      },
      web3: {
        activeNet: {
          id: 1,
        },
      },
      user: {
        settings: {
          fiatCurrency: 'USD',
        },
      },
      price: {
        price: 10,
      },
      connectionStatus: {
        isSyncing: true,
      },
    },

    modules: {
      transactions: {
        namespaced: true,
        getters: {
          allowedToSendAddresses: jest.fn(),
        },
      },

      tokens: {
        namespaced: true,
        getters: tokensGetters,
      },

      gasPrice: {
        namespaced: true,
        actions: gasPriceActions,
      },
    },

    mutations: {
      testActiveNet(state, net) {
        state.web3.activeNet = net;
      },
    },
  };

  beforeEach(() => {
    jest.clearAllMocks();

    store = new Vuex.Store(Object.assign({}, defaultStore));

    wrapperFactory = wrapShallowMountFactory(TransactionForm, {
      store,
      localVue,
      sync: false,
      provide: () => ({
        $validator: new VeeValidate.Validator(),
      }),
      propsData: {
        transaction: transactionProp,
      },
    });
    wrapper = wrapperFactory();
  });

  describe('render', () => {
    beforeEach(() => {
      wrapper.vm.estimateGasCost = jest.fn();
    });

    it('should render without error and resolved address by default', () => {
      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.find('.help.is-danger').exists()).toBe(false);
      expect(wrapper.find('.help.ellipsis').exists()).toBe(false);
    });

    it('should render message if ens address resolved', () => {
      wrapper = wrapperFactory({
        store,
        localVue,
        provide: () => ({
          $validator: new VeeValidate.Validator(),
        }),
        propsData: {
          transaction: {
            ...transactionProp,
            to: address,
          },
        },
        computed: {
          isEnsTransaction: true,
        },
      });

      wrapper.setData({
        ensError: null,
        isEnsAddressLoading: false,
      });

      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.find('.help.is-danger').exists()).toBe(false);
      expect(wrapper.find('.help.ellipsis').exists()).toBe(true);
      expect(wrapper.find('.help.ellipsis').text()).toBe(
        `Resolved ENS address: ${address}`,
      );
    });

    it('should render error on ens error', async () => {
      expect.assertions(4);

      wrapper = wrapperFactory({
        computed: {
          isEnsTransaction: true,
        },
      });
      wrapper.setData({
        ensError: 'foo',
        isEnsAddressLoading: false,
      });

      wrapper.vm.$forceUpdate();
      await wrapper.vm.$nextTick();

      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.find('.help.is-danger').exists()).toBe(true);
      expect(wrapper.find('.help.ellipsis').exists()).toBe(false);
      expect(wrapper.find('.help.is-danger').text()).toBe('foo');
    });

    it('should render priority options if gas price is not null', () => {
      wrapper.setData({
        prices: {
          low: 1,
          medium: 2,
          high: 3,
        },
      });

      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.find('transaction-priority-options-stub').exists()).toBe(
        true,
      );
    });
  });

  describe('behavior', () => {
    describe('methods', () => {
      it('should load gas price on mount', () => {
        expect(gasPriceActions.getGasPrice).toBeCalledTimes(1);
      });

      it('should merge transaction with given advanced options', () => {
        wrapper.vm.handleAdvancedChange({
          foo: 'bar',
        });

        expect(wrapper.vm.transaction).toEqual({
          ...transactionProp,
          foo: 'bar',
        });
      });

      describe('changeTokenInfo', () => {
        const tokenBySymbolGetterMock = jest.fn().mockReturnValue(tokens[0]);

        beforeEach(() => {
          store = new Vuex.Store(
            set(Object.assign({}, defaultStore), 'modules.tokens.getters', {
              ...tokensGetters,
              currentAccountTokenBySymbol: () => tokenBySymbolGetterMock,
            }),
          );
          wrapper = wrapperFactory({
            store,
            localVue,
            provide: () => ({
              $validator: new VeeValidate.Validator(),
            }),
            propsData: {
              transaction: transactionProp,
            },
          });
        });

        it('should change token info if it is not empty', () => {
          wrapper.vm.changeTokenInfo('SCDT');

          expect(tokenBySymbolGetterMock).toBeCalledWith('SCDT');
          expect(wrapper.vm.transaction.tokenInfo).toEqual(tokens[0]);
        });

        it('should not change token info if it is empty', () => {
          wrapper.vm.changeTokenInfo(null);

          expect(tokenBySymbolGetterMock).not.toBeCalled();
          expect(wrapper.vm.transaction.tokenInfo).toBe(null);
        });
      });

      describe('resolveEnsAddress', () => {
        it('shound resolve address and return it without error', async () => {
          expect.assertions(2);

          const res = await wrapper.vm.resolveEnsAddress();

          expect(res).toBe(address);
          expect(wrapper.vm.ensError).toBe(null);
        });

        it('shound handle error, print it and return empty string', async () => {
          expect.assertions(2);

          wrapper.setData({
            address: 'hello.eth',
          });

          wrapper.vm.$forceUpdate();
          await wrapper.vm.$nextTick();

          ENSResolver.getAddress.mockRejectedValueOnce(new Error());

          const res = await wrapper.vm.resolveEnsAddress();

          expect(res).toBe('');
          expect(wrapper.vm.ensError).toBe(
            'ENS hello.eth can not be resolved.',
          );
        });
      });

      describe('estimateGasCost', () => {
        beforeAll(() => {
          Transaction.getGasFullPrice = jest.fn();
          Transaction.isToContract = jest.fn();
        });

        it('should estimate gas cost for transaction', async () => {
          expect.assertions(2);

          Transaction.getGasFullPrice.mockResolvedValueOnce(1);

          await wrapper.vm.estimateGasCost();

          expect(Transaction.getGasFullPrice).toBeCalledWith(transactionProp);
          expect(wrapper.vm.estimatedGasCost).toBe(1);
        });

        it('should handle error', async () => {
          expect.assertions(3);

          const error = new Error();
          const previousEstimatedGas = wrapper.vm.estimatedGasCost;
          const previousEnsError = wrapper.vm.ensError;

          Transaction.getGasFullPrice.mockRejectedValueOnce(error);

          await wrapper.vm.estimateGasCost();

          expect(wrapper.vm.estimatedGasCost).toBe(previousEstimatedGas);
          expect(wrapper.vm.ensError).toBe(previousEnsError);
          expect(Transaction.isToContract).toBeCalledWith(
            wrapper.vm.transaction,
          );
        });

        it('should set ens error if failing transaction is contract and error contains always failing transaction', async () => {
          expect.assertions(3);

          const error = new Error('always failing transaction');
          const previousEstimatedGas = wrapper.vm.estimatedGasCost;

          Transaction.getGasFullPrice.mockRejectedValueOnce(error);
          Transaction.isToContract.mockResolvedValueOnce(false);

          await wrapper.vm.estimateGasCost();

          expect(wrapper.vm.estimatedGasCost).toBe(previousEstimatedGas);
          expect(wrapper.vm.ensError).toBe(
            'Transaction will always fail, try other address.',
          );
          expect(Transaction.isToContract).toBeCalledWith(
            wrapper.vm.transaction,
          );
        });
      });

      describe('loadGasPrice', () => {
        it('should request gas price and set transaction price as medium price', async () => {
          expect.assertions(1);

          await wrapper.vm.loadGasPrice();

          expect(wrapper.vm.transaction.gasPrice).toBe(
            gasPrice.medium.toString(),
          );
        });
      });
    });

    describe('watchers', () => {
      beforeEach(() => {
        jest.spyOn(wrapper.vm, 'debouncedGasCostEstimation');
        jest.spyOn(wrapper.vm, 'resolveEnsAddress');
      });

      it('should request gas cost on changing token info', async () => {
        expect.assertions(1);

        wrapper.setProps({
          transaction: {
            ...transactionProp,
            tokenInfo: {},
          },
        });

        await wrapper.vm.$nextTick();

        expect(wrapper.vm.debouncedGasCostEstimation).toBeCalledTimes(1);
      });

      it('should set address as empty string if transaction to is not exist', () => {
        wrapper.setProps({
          transaction: {
            ...transactionProp,
            to: '',
          },
        });

        expect(wrapper.vm.address).toBe('');
      });

      it('should resolve esn address and request gas cost on changing address to ens', async () => {
        expect.assertions(2);

        wrapper = wrapperFactory({
          computed: {
            isEnsTransaction: true,
          },
        });

        jest.spyOn(wrapper.vm, 'debouncedGasCostEstimation');
        jest.spyOn(wrapper.vm, 'resolveEnsAddress');

        ENSResolver.getAddress.mockResolvedValue(address);

        wrapper.setData({
          address: 'hello.eth',
        });

        await global.flushPromises();

        expect(wrapper.vm.resolveEnsAddress).toBeCalledTimes(1);
        expect(wrapper.vm.debouncedGasCostEstimation).toBeCalledTimes(1);
      });

      it('should request gas cost and set address to transaction on changing address', async () => {
        expect.assertions(2);

        wrapper.setData({
          address,
        });

        await wrapper.vm.$nextTick();

        expect(wrapper.vm.transaction.to).toBe(address);
        expect(wrapper.vm.debouncedGasCostEstimation).toBeCalledTimes(1);
      });

      it('should resolve ens address on change active network if current address is ens', async () => {
        expect.assertions(1);

        wrapper = wrapperFactory({
          computed: {
            isEnsTransaction: true,
          },
        });

        jest.spyOn(wrapper.vm, 'resolveEnsAddress');

        store.commit('testActiveNet', { id: 2 });

        await global.flushPromises();

        expect(wrapper.vm.resolveEnsAddress).toBeCalledTimes(1);
      });

      it('should not resolve ens address on change active network if network id was not changed', async () => {
        expect.assertions(1);

        // revert default value
        store.commit('testActiveNet', { id: 1 });

        wrapper = wrapperFactory({
          computed: {
            isEnsTransaction: true,
          },
        });

        jest.spyOn(wrapper.vm, 'resolveEnsAddress');

        store.commit('testActiveNet', { id: 1 });

        await global.flushPromises();

        expect(wrapper.vm.resolveEnsAddress).not.toBeCalled();
      });
    });
  });
});
