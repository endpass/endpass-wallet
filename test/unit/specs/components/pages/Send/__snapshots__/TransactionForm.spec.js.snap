// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Send – TransactionForm render should render error on ens error 1`] = `
<v-form id="sendEther" data-test="transaction-send-form" v-on:submit="bound handleFormSubmit">
    <div class="field">
        <label class="label">
            To
        </label>
        <account-chooser creatable="true" width="35" placeholder="0x... or ENS" data-test="transaction-address-select" v-model="address"></account-chooser>
        <!---->
        <p class="help is-danger">
            foo
        </p>
    </div>
    <transaction-amount-options current-token="[object Object]" fiat-currency="USD" eth-price="10" balance="1000000" estimated-gas-cost="0" disabled="true" active-net="[object Object]" v-on:change-token="bound changeTokenInfo" v-model="transaction.value"></transaction-amount-options>
    <transaction-priority-options v-model="transaction.gasPrice" prices="[object Object]"></transaction-priority-options>
    <transaction-advanced-options v-on:change="bound handleAdvancedChange"></transaction-advanced-options>
    <div class="field is-horizontal">
        <div class="field-label"></div>
        <div class="field-body">
            <v-button disabled="true" class-name="is-success is-medium is-cta" data-test="transaction-send-button">
                Send
            </v-button>
        </div>
    </div>
</v-form>
`;

exports[`Send – TransactionForm render should render message if ens address resolved 1`] = `
<v-form id="sendEther" data-test="transaction-send-form" v-on:submit="bound handleFormSubmit">
    <div class="field">
        <label class="label">
            To
        </label>
        <account-chooser creatable="true" width="35" placeholder="0x... or ENS" data-test="transaction-address-select" v-model="address"></account-chooser>
        <p class="help ellipsis">
            Resolved ENS address: 0xB14Ab53E38DA1C172f877DBC6d65e4a1B0474C3c
        </p>
        <!---->
    </div>
    <transaction-amount-options current-token="[object Object]" fiat-currency="USD" eth-price="10" balance="1000000" estimated-gas-cost="0" disabled="true" active-net="[object Object]" v-on:change-token="bound changeTokenInfo" v-model="transaction.value"></transaction-amount-options>
    <transaction-priority-options is-loading="true" v-model="transaction.gasPrice"></transaction-priority-options>
    <transaction-advanced-options v-on:change="bound handleAdvancedChange"></transaction-advanced-options>
    <div class="field is-horizontal">
        <div class="field-label"></div>
        <div class="field-body">
            <v-button disabled="true" class-name="is-success is-medium is-cta" data-test="transaction-send-button">
                Send
            </v-button>
        </div>
    </div>
</v-form>
`;

exports[`Send – TransactionForm render should render without error and resolved address by default 1`] = `
<v-form id="sendEther" data-test="transaction-send-form" v-on:submit="bound handleFormSubmit">
    <div class="field">
        <label class="label">
            To
        </label>
        <account-chooser creatable="true" width="35" placeholder="0x... or ENS" data-test="transaction-address-select" v-model="address"></account-chooser>
        <!---->
        <!---->
    </div>
    <transaction-amount-options current-token="[object Object]" fiat-currency="USD" eth-price="10" balance="1000000" estimated-gas-cost="0" disabled="true" active-net="[object Object]" v-on:change-token="bound changeTokenInfo" v-model="transaction.value"></transaction-amount-options>
    <transaction-priority-options v-model="transaction.gasPrice" prices="[object Object]"></transaction-priority-options>
    <transaction-advanced-options v-on:change="bound handleAdvancedChange"></transaction-advanced-options>
    <div class="field is-horizontal">
        <div class="field-label"></div>
        <div class="field-body">
            <v-button disabled="true" class-name="is-success is-medium is-cta" data-test="transaction-send-button">
                Send
            </v-button>
        </div>
    </div>
</v-form>
`;
